### PR TITLE
008 Adding a Login system

### PR DESCRIPTION
**Pull request by sdagn1**
- Added functionality to login to the system
- Created the Role Table to include roles for both User and Admin
- Created the User table which includes an email, hashed password, randomly generated salt, and role type
- Added functions to validate a login request by hashing the password after combining with the salt from the database using PBKDF2

### CODE REVIEW

**Review by Adnanhoq — APPROVED** [review]

**Review by Adnanhoq — APPROVED** [review]

**Review by AZZohaib — APPROVED** [review]

File: src/main/java/org/example/daos/AuthDao.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            return null;
+        }
+    }
+
+    public User validateUser(final LoginRequest loginRequest) throws
+            SQLException {
+
+        try (Connection connection = DatabaseConnector.getConnection()) {
+            String query = "SELECT Email, Password, Salt, RoleID FROM `User`"
+                    +
```
Comment (JennicaMistry) [inline src/main/java/org/example/daos/AuthDao.java:40]: This query is returning more fields than needed, so it's hurting the polar bears

File: src/main/java/org/example/validators/LoginValidator.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import org.example.exceptions.InvalidException;
+import org.example.models.LoginRequest;
+
+import java.util.regex.Pattern;
+import java.util.regex.Matcher;
+
+public class LoginValidator {
+    private final int emailLength = 255;
+    private final int passwordHashLength = 128;
+    private final String emailCheck = "^(?=.{1,64}@)[A-Za-z0-9_-]+"
```
Comment (JennicaMistry) [inline src/main/java/org/example/validators/LoginValidator.java]: Be consistent with your formatting

**Review by JennicaMistry — CHANGES_REQUESTED** [review]
There are no unit and integration tests in this PR.

File: src/main/java/org/example/daos/AuthDao.java (hunk 1 above)
Comment (sdagn1) [reply src/main/java/org/example/daos/AuthDao.java:40]: This has been corrected and committed. The fix still returns a user object, but sets the unnecessary String fields (password/salt) to the empty string of the Java object.

File: src/main/java/org/example/validators/LoginValidator.java (hunk 2 above)
Comment (sdagn1) [reply src/main/java/org/example/validators/LoginValidator.java]: This has been corrected and committed.

File: src/test/java/org/example/service/AuthServiceTests.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    LoginRequest validLoginRequest = new LoginRequest(
+            "admin@kainos.com",
+            "dFupuh0ITKJyNJEEyHSWwHRd5Mnhm65I+WfTUZWbOh8I+akEWJ6VDsFsWetFCdA1EQP+Mx193tVd7MudxjAMoBYqusV78vypAzv/TDfKYVmZCVsco8q7hskFtsWnKVCf"
+
+    );
+
+    User user = new User(
+            "admin@kainos.com",
+            "wlSNgEn5dCBM59jnbeH+txKWn36Vt6QScELcAa5ZBNduqSY16JAl2hqeGsZrmpG0kdb9+ILMoCJVB3er8ZoCJI9o26IM83UfnJtTT3p7cRgOUxsU0iMHgkI9KdQpDim6",
+            "salt",
```
Comment (JennicaMistry) [inline src/test/java/org/example/service/AuthServiceTests.java]: Missing test cases when the Dao returns null (one for each Dao method)

File: src/test/java/org/example/service/AuthServiceTests.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+        assertThrows(SQLException.class,
+                () -> authService.login(loginRequest));
+    }
+
+    @Test
+    void login_shouldReturnJwtToken_whenDaoReturnsUser() throws SQLException,
+            NoSuchAlgorithmException, InvalidKeySpecException, InvalidException {
+        Mockito.when(authDao.getUser(loginRequest)).thenReturn(user);
+        System.out.println(validLoginRequest.getPassword());
```
Comment (JennicaMistry) [inline src/test/java/org/example/service/AuthServiceTests.java]: Remove the System.out.println()

File: src/test/java/org/example/validator/LoginValidatorTests.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    LoginValidator loginValidator = new LoginValidator();
+
+    @Test
+    public void validateLogin_shouldNotThrowError_whenValidLogin() {
+        LoginRequest loginRequest = new LoginRequest(
+                "admin@kainos.com",
+                "wlSNgEn5dCBM59jnbeH+txKWn36Vt6QScELcAa5ZBNduqSY16JAl2hqeGsZrmpG0kdb9+ILMoCJVB3er8ZoCJI9o26IM83UfnJtTT3p7cRgOUxsU0iMHgkI9KdQpDim6"
+        );
+
+        assertDoesNotThrow(() -> loginValidator.validateLogin(loginRequest));
```
Comment (JennicaMistry) [inline src/test/java/org/example/validator/LoginValidatorTests.java]: We would like you to change these tests to parameterised tests - this is a new thing we have not shown you before. Below is an example with the function at the top and the parameterised test for it below. 
``` 
public static boolean isPerfectSquare(int num) {
    if (num < 0) {
        return false;
    }
    int sqrt = (int) Math.sqrt(num); // Calculate the square root of the number
    return sqrt * sqrt == num; // Check if the square of the square root equals the original number
} 
```

```
@ParameterizedTest
@CsvSource({"16, true",
            "14, false",
            "25, true",
            "0, true",
            "1, true",
            "-4, false",
            "20, false"})
public void testIsPerfectSquare(int input, boolean expected) {
    assertEquals(expected, employeeService.isPerfectSquare(input));
}
```

File: src/main/java/org/example/daos/AuthDao.java (hunk 1 above)
Comment (JennicaMistry) [reply src/main/java/org/example/daos/AuthDao.java:40]: You are still pulling back all the fields, and are not using your new constructor. You are still hurting the polar bears :(

**Review by JennicaMistry — CHANGES_REQUESTED** [review]

File: src/test/java/org/example/service/AuthServiceTests.java (hunk 4 above)
Comment (sdagn1) [reply src/test/java/org/example/service/AuthServiceTests.java]: This has been corrected and committed.

File: src/test/java/org/example/service/AuthServiceTests.java (hunk 3 above)
Comment (sdagn1) [reply src/test/java/org/example/service/AuthServiceTests.java]: This has been corrected, tested, and committed.

**Comment by JennicaMistry** [comment]
Also, the tests need to be added to the github action, either this can be done in this ticket or put into a separate ticket. Please decide among yourselves.

File: src/test/java/org/example/validator/LoginValidatorTests.java (hunk 5 above)
Comment (sdagn1) [reply src/test/java/org/example/validator/LoginValidatorTests.java]: This has been added, tested, and committed.

**Review by JennicaMistry — APPROVED** [review]
